File: mobile/lib/models/buyer.dart
class Buyer {
  const Buyer({
    required this.id,
    required this.name,
    required this.address,
    required this.phone,
    required this.gstin,
    required this.state,
    required this.stateCode,
    required this.isActive,
    required this.pendingPayable,
    this.whatsappNumber,
  });

  final String id;
  final String name;
  final String address;
  final String? phone;
  final String? gstin;
  final String? state;
  final String? stateCode;
  final bool isActive;
  final double pendingPayable;
  final String? whatsappNumber;

  Buyer copyWith({
    String? id,
    String? name,
    String? address,
    String? phone,
    String? gstin,
    String? state,
    String? stateCode,
    bool? isActive,
    double? pendingPayable,
    String? whatsappNumber,
  }) {
    return Buyer(
      id: id ?? this.id,
      name: name ?? this.name,
      address: address ?? this.address,
      phone: phone ?? this.phone,
      gstin: gstin ?? this.gstin,
      state: state ?? this.state,
      stateCode: stateCode ?? this.stateCode,
      isActive: isActive ?? this.isActive,
      pendingPayable: pendingPayable ?? this.pendingPayable,
      whatsappNumber: whatsappNumber ?? this.whatsappNumber,
    );
  }

  factory Buyer.fromJson(Map<String, dynamic> json) {
    return Buyer(
      id: json['id'].toString(),
      name: json['name'] as String? ?? '',
      address: json['address'] as String? ?? '',
      phone: json['phone'] as String?,
      gstin: json['gstin'] as String?,
      state: json['state'] as String?,
      stateCode: json['state_code'] as String?,
      isActive: json['is_active'] as bool? ?? true,
      pendingPayable: _toDouble(json['pending_payable']),
      whatsappNumber: json['whatsapp_number'] as String?,
    );
  }

  static double _toDouble(Object? value) {
    if (value is num) {
      return value.toDouble();
    }
    return double.tryParse(value?.toString() ?? '') ?? 0;
  }
}

File: backend/app/schemas/buyer.py
import uuid
from datetime import datetime
from decimal import Decimal, InvalidOperation
from typing import Literal

from pydantic import BaseModel, ConfigDict, field_validator

MONEY_QUANT = Decimal("0.01")
MAX_MONEY = Decimal("999999999999.99")


def validate_money_amount(value: Decimal) -> Decimal:
    if value <= 0:
        raise ValueError("amount must be greater than zero")
    if value > MAX_MONEY:
        raise ValueError("amount exceeds maximum supported value")
    try:
        if value != value.quantize(MONEY_QUANT):
            raise ValueError("amount must have at most two decimal places")
    except InvalidOperation as exc:
        raise ValueError("amount must have at most two decimal places") from exc
    return value


class BuyerCreateRequest(BaseModel):
    name: str
    address: str = ""
    phone: str | None = None
    gstin: str | None = None
    state: str | None = None
    state_code: str | None = None


class BuyerResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    name: str
    address: str
    phone: str | None
    gstin: str | None
    state: str | None
    state_code: str | None
    is_active: bool
    pending_payable: str = "0.00"


class BuyerLedgerEntryRequest(BaseModel):
    request_id: uuid.UUID
    amount: Decimal
    occurred_at: datetime
    notes: str | None = None

    @field_validator("amount")
    @classmethod
    def validate_positive_amount(cls, value: Decimal) -> Decimal:
        return validate_money_amount(value)

    @field_validator("occurred_at")
    @classmethod
    def validate_timezone_aware_datetime(cls, value: datetime) -> datetime:
        if value.tzinfo is None or value.utcoffset() is None:
            raise ValueError("occurred_at must include timezone information")
        return value


class BuyerAdjustmentRequest(BuyerLedgerEntryRequest):
    direction: Literal["INCREASE", "DECREASE"]


class BuyerTransactionResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    entry_type: str
    amount: Decimal
    occurred_at: datetime
    notes: str | None


class BuyerLedgerResponse(BaseModel):
    buyer: BuyerResponse
    transactions: list[BuyerTransactionResponse]
